### DISCUSSION TITLE
menu nav bar

### DISCUSSION THREAD

**Discussion by flosigime** [sqlpage/SQLPage#91] [Q&A]
Hello everybody, 

In the documentation of the shell component, we have this example : 

``` '["index","blog","components","functions"]' as menu_item ```

I'm trying to build a menu but, if i do the same, in my case, the result on our page is this : 

```html
<li class="nav-item">
<a class="nav-link text-capitalize" href="[[&quot;index&quot;,&quot;blog&quot;,&quot;components&quot;,&quot;functions&quot;].sql](view-source:http://localhost:8080/[%22index%22,%22blog%22,%22components%22,%22functions%22].sql)">["index","blog","components","functions"]</a>
</li>
```

I'd like to find a solution to build my menu easier than repeating on each page : 

```sql
'index' as menu_item,
'blog' as menu_item,
'cimponents' as menu_item,
'functions' as menu_item,
```

thanks for your help

**Comment by lovasoa** [comment]
Hi Florent ! I hope you had a nice summer !

  The display of `'["index","blog","components","functions"]' as menu_item` is a problem on the documentation site. There is no good cross-database way to define an inline json object in SQL.

   Which database did you settle on ? SQLite or Postgres ?
- In SQLite, there is no native json object, there are only strings, so you are stuck with the repeated ` as menu_item` if you want to define an inline json object.
- In Postgres, you can use a native json object with `select '["index","blog","components","functions"]'::json as menu_item`


Independently of that, the best way to factorize your menus to avoid having to repeat them on every page is to store them directly in your database (create a small `sqlpage_shell` table with a single row), and then do a simple

```sql
select * from sqlpage_shell
```

at the top of all your pages.

**Comment by flosigime** [comment]
Hi Ophir. Yes, really not bad, thanks. I hope the same for you, We saw SQLpage made a lot of progress during this time :-)

Here, we're still on Sqlite for the moment, for field recording. I could send you what we did yet, There is probably a lot of things to improve...

I saw in your example the idea with the sqlpage_shell table. Everything is not totally clear for me. 
First : we have to create the table in the sqlite database have'nt we ? It seems we can't use migration to do that if we have a database yet. 
Second : the sqlpage_shell table contains a single row, with a field menu_item like here : 

```sql
INSERT INTO shell(title, description, icon, link, menu_item, font, lang)
VALUES ('Corporate Conundrum',  'Unleash your inner executive in this thrilling board game of corporate espionage. Make the right choices to lead your company to success!', 'affiliate', '/', 'New Game', 'Libre Baskerville', 'en-US'); 
```

But, if we want to have more than one menu item ? 

Thanks for your answers

**Comment by lovasoa** [comment]
> Here, we're still on Sqlite for the moment, for field recording. I could send you what we did yet, There is probably a lot of things to improve...

Yes, please reach out by email, I'll be glad to continue the discussions we started earlier.


>  First : we have to create the table in the sqlite database have'nt we ? It seems we can't use migration to do that if we have a database yet.

I'm not sure I understand where the problem is. Yes, you have to create the table in your database file. You can do that through SQLPage using a migration, or through any other tool you like. Why couldn't use a migration if you already have a database ?

>  Second : the sqlpage_shell table contains a single row, with a field menu_item like here 

Yes, creating a table avoids the copy & paste between all your sql files, but it does not solve your problem with multiple menu items on databases that do not support native json objects such as SQLite. I think I'll make future versions of SQLPage automatically interpret the menu items as json. That'll make your original syntax (`'["a","b","c"]' as menu_item`) valid and make it as easy to use multiple menu items in sqlite as it is in postgres.

**Reply by lovasoa** [reply]
I'll publish it as a part of 0.12.0

**Comment by flosigime** [comment]
> I'm not sure I understand where the problem is. Yes, you have to create the table in your database file. You can do that through SQLPage using a migration, or through any other tool you like. Why couldn't use a migration if you already have a database ?

It's for not disturbing our collegues. The sqlite database, after field recording, is imported in Qgis when we're back in our office. Using the ability of SqlPage to use the migration folder with a "temporary" table will prevent creating a table in the sqlite database only for Sqlpage

**Reply by lovasoa** [reply]
If you add a `CREATE TABLE` statement to a migration, then the table will be saved to the sqlite database file the same way as if you had created it with another tool.

[SQLite temporary tables](https://www.sqlite.org/lang_createtable.html) are unrelated to sqlpage migrations, and they are probably not what you want here. They are specific to one database connection, and you want a table that will persist between connections.
